plutus-tx/src/PlutusTx/Arbitrary.hs

```diff
@@ -0,0 +1,164 @@
+{-# LANGUAGE DerivingStrategies   #-}
+{-# LANGUAGE DerivingVia          #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE TupleSections        #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module PlutusTx.Arbitrary where
+
+import qualified Data.List                   as List
+
+import           Test.QuickCheck             (Arbitrary (..), Gen, Positive (Positive, getPositive), genericShrink,
+                                              getOrdered, oneof, resize, sized, suchThat)
+import           Test.QuickCheck.Instances   ()
+
+import           PlutusTx                    (Data (B, Constr, I, List, Map))
+import           PlutusTx.AssocMap           (Map)
+import qualified PlutusTx.AssocMap           as AssocMap (fromList, toList)
+import           PlutusTx.Bimap              (Bimap)
+import qualified PlutusTx.Bimap              as Bimap
+import           PlutusTx.Builtins
+import           PlutusTx.Builtins.Internal
+import           PlutusTx.NatRatio.Internal  (NatRatio (NatRatio))
+import           PlutusTx.Natural.Internal   (Natural (Natural))
+import           PlutusTx.NonEmpty           (NonEmpty ((:|)))
+import qualified PlutusTx.Ord                as Ord
+import qualified PlutusTx.Prelude            as PlutusPrelude
+import           PlutusTx.Ratio              (denominator, numerator, (%))
+import qualified PlutusTx.Ratio              as Ratio (Rational, fromGHC, toGHC)
+import           PlutusTx.Set                (Set)
+import qualified PlutusTx.Set                as Set
+import           PlutusTx.UniqueMap.Internal (UniqueMap (UniqueMap))
+
+--------------------------------------------------------------------------------
+-- Special cases
+
+instance Arbitrary BuiltinByteString where
+  arbitrary = BuiltinByteString <$> resize 64 arbitrary
+  shrink (BuiltinByteString x) = BuiltinByteString <$> shrink x
+
+instance Arbitrary Data where
+  -- ToJSON / FromJSON instances of BuiltinData implemented with serialise / deserialise functions from Codec.Serialise.
+  -- In case the length of ByteStrings is more than 64 bytes, deserialization error occurs.
+  --
+  -- Example:
+  -- B "\RS\194w\255$u\144-\191\205u\191\183\&4\232\190\225\249\"}S+\217\ENQf\196\SUB\SI\191\169\149^\DC41n%\252\&1\245wr&\161|OX3\170h7\153\210\225\177\228\233\207\231#\DC1[\191\242\148\210\US\167"
+  -- Failed! Exception: 'DeserialiseFailure 69 "ByteString exceeds 64 bytes"'
+  --
+  -- It goes from the implementation of PlutusCore.Data.decodeBoundedBytes:
+  --
+  -- @
+  -- decodeBoundedBytes :: Decoder s Data
+  -- decodeBoundedBytes =  do
+  -- b <- CBOR.decodeBytes
+  -- if BS.length b <= 64
+  --  then pure $ B b
+  --  else fail $ "ByteString exceeds 64 bytes"
+  -- @
+  arbitrary = sized $ oneof . dataGenList
```

Comment:
There's already some generation code for `Data` in the `plutus-tx` tests, we could take some of that.